src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/InlineCompletion/InlineCompletionEndPoint.cs

```diff
@@ -0,0 +1,238 @@
+        {
+            return null;
+        }
+
+        var sourceText = await document.GetTextAsync();
+        var linePosition = new LinePosition(request.Position.Line, request.Position.Character);
+        var hostDocumentIndex = sourceText.Lines.GetPosition(linePosition);
+
+        var languageKind = _documentMappingService.GetLanguageKind(codeDocument, hostDocumentIndex);
+
+        // Map to the location in the C# document.
+        if (languageKind != RazorLanguageKind.CSharp ||
+            !_documentMappingService.TryMapToProjectedDocumentPosition(codeDocument, hostDocumentIndex, out var projectedPosition, out _))
+        {
+            _logger.LogInformation($"Unsupported location for {request.TextDocument.Uri}.");
+            return null;
+        }
+
+        request.Position = projectedPosition;
+        var response = await _languageServer.SendRequestAsync(LanguageServerConstants.RazorInlineCompletionEndpoint, request).ConfigureAwait(false);
+        var list = await response.Returning<InlineCompletionList>(cancellationToken);
+        if (list == null)
+        {
+            _logger.LogInformation($"Did not get any inline completions from delegation.");
+            return null;
+        }
+
+        var items = new List<InlineCompletionItem>();
+        var csharpDocOptions = codeDocument.GetCSharpDocument();
+        foreach (var item in list.Items)
+        {
+            var containsSnippet = item.TextFormat == InsertTextFormat.Snippet;
+            var range = item.Range ?? new Range { Start = projectedPosition, End = projectedPosition };
+
+            if (!_documentMappingService.TryMapFromProjectedDocumentRange(codeDocument, range, out var rangeInRazorDoc))
+            {
+                _logger.LogWarning($"Could not remap projected range {range} to razor document");
+                continue;
+            }
+
+            using var formattingContext = FormattingContext.Create(request.TextDocument.Uri, document, codeDocument, s_defaultFormattingOptions, _adhocWorkspaceFactory, isFormatOnType: true, automaticallyAddUsings: false);
+            if (!TryGetSnippetWithAdjustedIndentation(formattingContext, item.Text, hostDocumentIndex, out var newSnippetText))
+            {
+                continue;
+            }
+
+            var remappedItem = new InlineCompletionItem
+            {
+                Command = item.Command,
+                Range = rangeInRazorDoc,
+                Text = newSnippetText.ToString(),
+                TextFormat = item.TextFormat,
+            };
+            items.Add(remappedItem);
+        }
+
+        _logger.LogInformation($"Returning items.");
+        return new InlineCompletionList
+        {
+            Items = items.ToArray()
+        };
+    }
+
+    private static bool TryGetSnippetWithAdjustedIndentation(FormattingContext formattingContext, string snippetText, int hostDocumentIndex, [NotNullWhen(true)] out string? newSnippetText)
```

Comment:
While redoing this code I found a bug in formatting (both in the old version and this version) related to tabstops in LSP snippets.  This ended up being by far the easiest way to get a properly indented snippet.   

Basically, formatting text edits via the razor formatter causes the C# formatter to run against a document with the LSP snippet inserted.  Unfortunately snippets have placeholder fields which cause syntax errors.  Previously caret locations were wrapped by comment blocks (`/*$0*/`) but fields were not.  Since fields in LSP snippets are demarcated by invalid C# syntax, the C# formatter will not format this correctly.
E.g.
```
if (${1:true})
{
}
```
is completely invalid C#.

Two potential options here
1.  Similar to caret locations, wrap fields in comment blocks.  This kind of works, but wrapping fields still leaves us with syntax errors since the fields represent actual code.  For example `if (/*${1:true}*/)` is invalid C# code which doesn't solve the problem of us formatting invalid C# and getting weird formatting results.  It does seem to recover slightly better, but not super great. 
2.  Truly the only way to run the C# formatter against the snippet is to literally replace all fields with the default values, run the formatter, and then somehow extract back out the new field locations.  This is essentially re-implementing the code I already wrote on the C# server side (and therefore running it twice), but also having to parse the LSP snippet to extract out the fields + integrate that with the extraordinarly complex razor formatting logic.
3.  So I instead chose the third option, which is to essentially ignore the razor formatter.  C# snippets should already be formatted appropriately according to the underlying document.  We just need to replace the C# indentation with the razor indentation.  This appears to work rather nicely, but as I am not that familiar with Razor formatting I may have missed something :).   Tagging @NTaylorMullen for thoughts here.
See gif
![razor_snippets_formatting](https://user-images.githubusercontent.com/5749229/151347614-90fde223-02b1-4b3f-bcbf-d273991b8772.gif)

Reply:
lmao I'm actually a bit surprised 3 works only because there are many portions in the document where the underlying indentation is relatively busted. Does doing `@for[TAB]` at the beginning of a line work as you'd expect?

@davidwengier would probably know better on if we can take this approach but my gut tells me we probably have to approach your option 2

Reply:
> Does doing @for[TAB] at the beginning of a line work as you'd expect?
Yup!
![image](https://user-images.githubusercontent.com/5749229/151423582-18e9470c-5a82-4fba-ae5c-8b8a865b206a.png)

if we go with 2, the C# server is basically doing nothing useful.  We might as well parse the xml snippet in razor, would be much easier to implement 😂

Reply:
Option 3 makes sense to me. I could see an argument for the editor sending a range formatting request when the snippet session is closed or something, but I don't think we need to try to do anything fancy here. It seems fair to assume the snippet is correctly formatted in and of itself (and I'd go further with that assumption.. see other comment below), and ensure we're inserting it in the right spot, but formatting it to our razor rules might not actually be desirable to the user anyway - some of them never format the document, and would like us to not format on type either.